plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,210 @@
+import com.mapbox.maps.plugin.gestures.gestures
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION
+import com.mapbox.maps.plugin.viewport.experimental.data.DefaultViewportTransitionOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.FollowingViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.OverviewViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransitionImpl
+import com.mapbox.maps.plugin.viewport.experimental.transition.ImmediateViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransition
+import java.lang.IllegalStateException
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStatusObservers = CopyOnWriteArrayList<ViewportStatusObserver>()
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+  private var currentCancelable: Cancelable? = null
+  private val transitions = HashMap<Pair<ViewportState?, ViewportState?>, ViewportTransition>()
+  private lateinit var delegateProvider: MapDelegateProvider
+  private lateinit var gesturesPlugin: GesturesPlugin
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        currentCancelable?.cancel()
+        currentCancelable = null
+        updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION)
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      // no-ops
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {
+      // no-ops
+    }
+  }
+
+  override val states: Set<ViewportState>
+    get() = internalStates.toSet()
+
+  override var status: ViewportStatus = ViewportStatus.State(null)
+    private set
+
+  private fun updateStatus(targetStatus: ViewportStatus, reason: String) {
+    val previousStatus = status
+    status = targetStatus
+    notifyStatusChanged(previousStatus, targetStatus, reason)
+  }
+
+  override fun transitionTo(targetState: ViewportState, completionListener: CompletionListener?) {
```

Comment:
since an overriding function cannot provide default arguments, we might want to add the override manually to the interface, for ease of use

Reply:
@LukasPaczos We have added the default value to the interface, verified that the overloads are working as expected in the test app in Koltin.
however it seems we can not use JvmOverloads annotation on the interface, for Java users, these overloads wouldn't be generated, do you suggest we add overload functions manually in the interface? How do you deal with these situations in Nav SDK?